ql/lib/codeql/kaleidoscope/Ast.qll

```diff
@@ -0,0 +1,410 @@
+private import codeql.kaleidoscope.ast.internal.Ast
+private import codeql.kaleidoscope.ast.internal.TreeSitter
+private import codeql.Locations
+
+class AstNode extends TAstNode {
```

Comment:
A pattern that I have started using is to define internal implementation classes, using `abstract` and dispatch, and then reference those classes in the public API, where all classes are final. I.e., something like
```
abstract class AstNodeImpl extends TAstNode {
  abstract AstNode getAChild(string name);

  AstNode getParent() { result.getAChild(_) = this }

  abstract string toString();

  abstract getAPrimaryQlClass();

  abstract Location getLocation();
}

abstract class ExpressionImpl extends AstNodeImpl, TExpression { }

abstract class BinaryOpExpressionImpl extends ExpressionImpl, TBinaryOpExpression { /* ... */ }
```
in `internal/Ast.qll` and then in this file
```
final class AstNode instanceof AstNodeImpl {
  AstNode getAChild(string name) { result = super.getAChild(name) }

  AstNode getParent() { result = super.getParent() }

  string toString() { result = super.toString() }

  string getAPrimaryQlClass() { result = "???" }

  Location getLocation() { result = super.getLocation() }
}

final class Expression extends AstNode instanceof ExpressionImpl { }

final class BinaryOpExpression extends Expression instanceof BinaryOpExpressionImpl { /* ... */ }
```